op-supervisor/supervisor/backend/rewinder/rewinder.go

```diff
@@ -0,0 +1,256 @@
+}
+
+// Rewinder is responsible for handling the rewinding of databases to the latest common ancestor between
+// the local databases and L2 node.
+type Rewinder struct {
+	log     log.Logger
+	emitter event.Emitter
+	l1Node  l1Node
+	db      rewinderDB
+}
+
+func New(log log.Logger, db rewinderDB, l1Node l1Node) *Rewinder {
+	return &Rewinder{
+		log:    log.New("component", "rewinder"),
+		db:     db,
+		l1Node: l1Node,
+	}
+}
+
+func (r *Rewinder) AttachEmitter(em event.Emitter) {
+	r.emitter = em
+}
+
+func (r *Rewinder) OnEvent(ev event.Event) bool {
+	switch x := ev.(type) {
+	case superevents.RewindL1Event:
+		r.handleRewindL1Event(x)
+		return true
+	case superevents.LocalDerivedEvent:
+		r.handleLocalDerivedEvent(x)
+		return true
+	default:
+		return false
+	}
+}
+
+// handleRewindL1Event iterates known chains and checks each one for a reorg
+// If a reorg is detected, it will rewind the chain to the latest common ancestor
+// between the local-safe head and the finalized head.
+func (r *Rewinder) handleRewindL1Event(ev superevents.RewindL1Event) {
+	for _, chainID := range r.db.DependencySet().Chains() {
+		if err := r.rewindL1ChainIfReorged(chainID, ev.IncomingBlock); err != nil {
+			r.log.Error("failed to rewind L1 data:", "chain", chainID, "err", err)
+		}
+	}
+}
+
+// handleLocalDerivedEvent checks if the newly derived block matches what we have in our unsafe DB
+// If it doesn't match, we need to rewind the logs DB to the common ancestor between
+// the LocalUnsafe head and the new LocalSafe block
+func (r *Rewinder) handleLocalDerivedEvent(ev superevents.LocalDerivedEvent) {
+	// Get the block at the derived height from our unsafe chain
+	newSafeHead := ev.Derived.Derived
+	unsafeVersion, err := r.db.FindSealedBlock(ev.ChainID, newSafeHead.Number)
+	if err != nil {
+		r.log.Error("failed to get unsafe block at derived height", "chain", ev.ChainID, "height", newSafeHead.Number, "err", err)
+		return
+	}
+
+	// If the block hashes match, our unsafe chain is still valid
+	if unsafeVersion.Hash == newSafeHead.Hash {
+		return
+	}
+
+	// Try rewinding the logs DB to the parent of the new safe head
+	// If it fails with a data conflict walk back through the chain
+	// until we find a common ancestor or reach the finalized block
+	finalized, err := r.db.Finalized(ev.ChainID)
+	if err != nil {
+		r.log.Error("failed to get finalized block", "chain", ev.ChainID, "err", err)
+		return
+	}
+	for height := int64(newSafeHead.Number - 1); height >= int64(finalized.Number); height-- {
+		// Get the block at this height
+		block, err := r.db.FindSealedBlock(ev.ChainID, uint64(height))
+		if err != nil {
+			r.log.Error("failed to get sealed block", "chain", ev.ChainID, "height", height, "err", err)
+			return
+		}
+
+		// Try to rewind and stop if it succeeds
+		err = r.db.RewindLogs(ev.ChainID, block)
+		if err == nil {
+			break
+		}
+
+		// If it failed with a data conflict try the next block
+		if errors.Is(err, types.ErrConflict) {
```

Comment:
I am not sure if you want to be using `FindSealedBlock` in this loop -- this will get the block data from the *Log Database*, which at this point in the process is the thing you're trying to reorg *out*. For that reason, the check for `ErrConflict` will never catch a conflict.

Rather, I expect you want to get the block reference from the *Local Safe DB*, which has the more correct hash, and will show you when the Rewind finds a matching head.

Reply:
Fixed and tests are all passing 🔥